### PR TITLE
[varLib] Support sparse masters

### PR DESCRIPTION
**Pull request by behdad**

### CODE REVIEW

**Comment by justvanrossum** [comment]
Great that this is getting done, thanks!

I have an idea that relates to this, and I wonder what you'd think.

Sometimes I've had a need to have masters/deltas for metrics (HVAR/GPOS) but _not_ for outlines. With your current work, it may become feasible to implement the following: if a (non-neutral) glyph _exists_, but has no outlines or components, then interpolation compatibility checks are bypassed, and only the advance width/height values are taken into consideration.

File: Lib/fontTools/varLib/models.py [hunk 1]
```diff
@@ -314,6 +345,10 @@ def getDeltas(self, masterValues):
 			out.append(delta)
 		return out
 
+	def getDeltasAndSupports(self, items):
+		model, items = getSubModel(items)
```
[the remark this answers is not in the document]
Comment (behdad) [reply Lib/fontTools/varLib/models.py]: Ouch.  yeah. Done.

**Comment by behdad** [comment]
Okay. First chunk of work is mostly done.  HVAR was tricky but done.  Only need to allow masters without GPOS and would be done for the first phase.  More complicated GPOS kerning work is next stage.

**Comment by behdad** [comment]
(Fixing the tests.)

**Comment by behdad** [comment]
This is ready as far as I'm concerned.

**Comment by behdad** [comment]
> Sometimes I've had a need to have masters/deltas for metrics (HVAR/GPOS) but _not_ for outlines. With your current work, it may become feasible to implement the following: if a (non-neutral) glyph _exists_, but has no outlines or components, then interpolation compatibility checks are bypassed, and only the advance width/height values are taken into consideration.

The more I think about this, the more it doesn't make sense to me, at least not if glyph advance is also ignored.  Setting different advance width without drawing the glyph sounds wrong to me.

**Comment by behdad** [comment]
I'm now going to dig deeper into GPOS, trying merging GPOS where a master misses a glyph.   Maybe that isn't a common use case and I shouldn't try.  Probably yes.  It's easy to do for anchors and possibly kerning, but if there's any other feature, it becomes nasty.  But maybe getting that much done would make it UFO-based fontmake workflow where there's no handwritten `features.fea` file work without any changes needed in the rest of the stack, and that's a good start?

**Comment by behdad** [comment]
Reached another milestone: sparse mark-base positioning is also merged now.

The way I tested this:

- Built UFO from Glyphs file for NotoSerifTamil:
```
$ fontmake -o ufo -g NotoSerifTamil-MM.glyphs
```

- Modified `master_ufo/NotoSerifTamil-Light.ufo/glyphs/contents.plist` and `master_ufo/NotoSerifTamil-Light.ufo/features.fea` and deleted `a-tamil` from them.

- Built interpolatable masters:
```
$ fontmake -o ttf-interpolatable -m master_ufo/NotoSerifTamil.designspace
```

- Built variable font using varLib:
```
$ fonttools varLib master_ufo/NotoSerifTamil.designspace
```

It builds for me now.

Next: find a glyph with mark-to-mark positioning, and implement that.  Then mark-to-ligature.

Next next: kerning; it's *almost* there.  Needs tiny amount of work.  Right now merger assumes missing kerning is 0.  Need to distinguish that from glyph-not-in-font.

**Comment by behdad** [comment]
> Next next: kerning; it's _almost_ there. Needs tiny amount of work. Right now merger assumes missing kerning is 0. Need to distinguish that from glyph-not-in-font.

That's a tricky issue.  One can argue that if kerning is missing it should not be assumed 0, and discard the master for that pair.  Might result in surprises, eg, if one doesn't kern a pair in Bold master because it wasn't needed, but is kerned in Black, then if we don't assume 0, the kern for Black will interpolate all the way to Regular, changing the desired kern at Bold.

However, not assuming 0 allows for having sparse masters with no kerning at all, or no kerning for particular glyphs...  which otherwise would have to.

**Comment by behdad** [comment]
MarkMarkPos implemented as well.

**Comment by behdad** [comment]
Okay, fixed.  Again, ready to merge.

**Comment by justvanrossum** [comment]
> The more I think about this, the more it doesn't make sense to me, at least not if glyph advance is also ignored. Setting different advance width without drawing the glyph sounds wrong to me.

That is the whole point, to have a delta for the advance and not the outline. It may make no sense to you, but I have actually needed this for a real project. The use case exists.

**Comment by behdad** [comment]
> That is the whole point, to have a delta for the advance and not the outline. It may make no sense to you, but I have actually needed this for a real project. The use case exists.

Did you really just needed to add more space just to the right side of the glyph shape, and not left side?  Because that's what happens if only advance width varies.  And this sounds like as quite an arbitrary limitation / need to me. (the one-sided-ness)

**Comment by justvanrossum** [comment]
> Did you really just needed to add more space just to the right side of the glyph shape, and not left side? Because that's what happens if only advance width varies. And this sounds like as quite an arbitrary limitation / need to me. (the one-sided-ness)

Yes, exactly. It was about tweaking the metrics of an intermediate master, so it would match the metrics of an existing static font perfectly. We stuck with pure interpolation for most outlines, as that was good enough, and metrics compatibility was more important than exact outline matching. And saving space was an issue. This solution gave us the best tradeoffs among all requirements.

**Comment by behdad** [comment]
> Yes, exactly. It was about tweaking the metrics of an intermediate master, so it would match the metrics of an existing static font perfectly. We stuck with pure interpolation for most outlines, as that was good enough, and metrics compatibility was more important than exact outline matching. And saving space was an issue. This solution gave us the best tradeoffs among all requirements.

Yeah okay sure, that is one legitimate use case :).

**Comment by justvanrossum** [comment]
> Yeah okay sure, that is one legitimate use case :).

I don't insist that varLib implements this (I managed ok with a post processing script), I was just thinking that perhaps with your current work it would be trivial to support. In which case it would be a win-win :)

**Comment by behdad** [comment]
> I don't insist that varLib implements this (I managed ok with a post processing script), I was just thinking that perhaps with your current work it would be trivial to support. In which case it would be a win-win :)

Oh I already did last night. https://github.com/fonttools/fonttools/pull/1368/commits/446ee2a4d011909eaa14fa5ccbd8c4d6df3329de

**Comment by justvanrossum** [comment]
> Advance width comes from hmtx/HVAR. You are thinking UFO.

I'm thinking phantom points. Shouldn't they stay in sync with HVAR?

**Comment by behdad** [comment]
> > Advance width comes from hmtx/HVAR. You are thinking UFO.
> 
> I'm thinking phantom points. Shouldn't they stay in sync with HVAR?

There's no way I can include phantom points but not the rest of the glyph.

**Comment by behdad** [comment]
But yeah, for that reason, maybe this is not a good idea.

The alternative is hard: to interpolate the glyph and insert it if it's missing.  I don't want to go that far.

**Comment by justvanrossum** [comment]
> There's no way I can include phantom points but not the rest of the glyph.

Uh, of course you can: you can have deltas for the phantom points and no deltas for the rest of the points.

**Comment by behdad** [comment]
> Uh, of course you can: you can have deltas for the phantom points and no deltas for the rest of the points.

Sometimes you can.  In general you can't.

Imagine a simple system with Regular, Bold, and Black masters.  There's two ways to model that.  Both result in same output normally (modulo rounding issues):

  1) Black can go from Regular all the way to Black; Bold will go from Regular up to Bold and down to Black.  Note how Bold master here will be "patching up" whatever Black master leaves as bold at the location of Bold master.

  2) Bold starts from Regular going up to Bold and going down to Black; Black master starting at Bold and going up to Black.

In the two equivalent systems above, in the first one, if you don't encode deltas for a glyph shape in Bold master, you get whatever else the system would do.  Ie. this system is equivalent to what you get if you deleted the Bold master.  The second solution however is not.  If you don't encode glyph shape in Bold master, you get the shape for Regular, which is not what you expect.

So, basically what you want, is to use the variation model without a particular master, encode glyph shapes, *then* add phantom-point deltas to fix up advance width to pre-defined values at particular points.  That's more than just a couple lines here or there, and needs more thought as to how that would interact with other masters.

**Comment by justvanrossum** [comment]
Got it, I see that it's more complicated than I thought. What you describe is if I recall correctly indeed more or less what I implemented in my post processing script. But it didn't need to cover all possibilities, as the font in question had a simple predictable model and only one axis.

**Comment by behdad** [comment]
Should I revert that commit then?  Because, as you point out, according to spec, HVAR and glyph phantom points must agree on the glyph width.

**Comment by justvanrossum** [comment]
> Should I revert that commit then?

I think so, yes.

**Comment by behdad** [comment]
If someone wants to give me a test case (smaller better) with kerning, I can look into that.  Right now the code assumes a kerning of 0 if a kerning pair doesn't appear in a master, even if the master does not contain one or two of those glyphs.
